hid.pyx

```diff
@@ -65,21 +66,23 @@ def enumerate(int vendor_id=0, int product_id=0):
     hid_free_enumeration(info)
     return res
 
-def hidapi_exit():
-    """Callback for when the script exits.
-
-    This prevents memory leaks in the hidapi C library.
-    Note that the counterpart, hid_init(), is not called explicitly. It will
-    be called internally when first needed.
-    """
-    hid_exit()
-
 cdef class device:
     """Device class.
 
     A device instance can be used to read from and write to a HID device.
     """
+
     cdef hid_device *_c_hid
+    cdef object __weakref__ # enable weak-reference support
+
+    def __cinit__(self):
+        """Initialize the device instance.
+
+        Sets up `self.close()` to be automatically called once the strong count
+        goes to zero.  This is necessary to prevent issues later in the
+        execution, when finalizing the HIDAPI library itself.
+        """
+        weakref.finalize(self, self.close)
```

Comment:
@jonasmalacofilho I'm wondering how this works because the [doc](https://docs.python.org/3/library/weakref.html#weakref.finalize) of `weakref.finalize(obj, func, /, *args, **kwargs)` say:

> Note: It is important to ensure that func, args and kwargs do not own any references to obj, either directly or indirectly, since otherwise obj will never be garbage collected. In particular, func should not be a bound method of obj.

Reply:
In my testing both the device was closed and `hid_exit()` was called, but it might have been just luck.

I'll take another look...